adbc.h

```diff
@@ -1095,6 +1258,112 @@ ADBC_EXPORT
 AdbcStatusCode AdbcConnectionRelease(struct AdbcConnection* connection,
                                      struct AdbcError* error);
 
+/// \brief Cancel the in-progress operation on a connection.
+///
+/// This can be called during AdbcConnectionGetObjects (or similar),
+/// or while consuming an ArrowArrayStream returned from such.
+/// Calling this function should make the other functions return
+/// ADBC_STATUS_CANCELLED (from ADBC functions) or ECANCELED (from
+/// methods of ArrowArrayStream).
+///
+/// This must always be thread-safe (other operations are not).
+///
+/// \since ADBC API revision 1.1.0
+/// \addtogroup adbc-1.1.0
+///
+/// \param[in] connection The connection to cancel.
+/// \param[out] error An optional location to return an error
+///   message if necessary.
+///
+/// \return ADBC_STATUS_INVALID_STATE if there is no operation to cancel.
+/// \return ADBC_STATUS_UNKNOWN if the operation could not be cancelled.
+ADBC_EXPORT
+AdbcStatusCode AdbcConnectionCancel(struct AdbcConnection* connection,
+                                    struct AdbcError* error);
+
+/// \brief Get a hierarchical view of system and user-defined functions.
+///
+/// The result is an Arrow dataset with the following schema:
+///
+/// | Field Name               | Field Type              |
+/// |--------------------------|-------------------------|
+/// | catalog_name             | utf8                    |
+/// | catalog_db_schemas       | list<DB_SCHEMA_SCHEMA>  |
+///
+/// DB_SCHEMA_SCHEMA is a Struct with fields:
+///
+/// | Field Name               | Field Type              |
+/// |--------------------------|-------------------------|
+/// | db_schema_name           | utf8                    |
+/// | db_schema_functions      | list<FUNCTION_SCHEMA>   |
+///
+/// FUNCTION_SCHEMA is a Struct with fields:
+///
+/// | Field Name               | Field Type              | Comments |
+/// |--------------------------|-------------------------|----------|
+/// | function_name            | utf8 not null           |          |
+/// | remarks                  | utf8                    | (1)      |
+/// | function_type            | int16                   | (2)      |
+/// | specific_name            | utf8                    | (3)      |
+/// | function_columns         | list<COLUMN_SCHEMA>     |          |
+///
+/// 1. Database-specific description of the function.
+/// 2. The kind of function.  Should be null if not known, 1 if the function
```

Comment:
Postgres has aggregate, window, procedure and standard functions. Not sure how all databases manage this but seems like we might want to have finer granularity than just the 1/2 proposed here?

Reply:
Also worth noting that while the proposal here has separate interfaces for functions versus procedures, postgres considers both to be a type of routine. Client tools like psql will list both via the `\df` shortcut

Reply:
Yes, this tried to adhere fairly closely to JDBC and so it's not very expressive. But I think so far it's mostly folly to try to unify these different definitions until we see some more concrete use cases.

Reply:
Out of curiosity where do you find the JDBC specification details? Having trouble finding a definitive authority on that via search

Reply:
Mostly just the API reference documentation; the specification itself hasn't been helpful to me (also, it seems it's released as revisions to previous specs instead of as a standalone spec, so e.g. even though [this](https://download.oracle.com/otn-pub/jcp/jdbc-4_3-mrel3-eval-spec/jdbc4.3-fr-spec.pdf?AuthParam=1687272585_864c20ce8d42bd6fd6806595083f3975) is "JDBC 4.3" it doesn't actually describe most things)